src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorSourceUpdater.cs

```diff
@@ -48,59 +56,332 @@ public async Task<IUpdaterResult> IsApplicableAsync(IUpgradeContext context, Imm
+            }
+
+            return new FileUpdaterResult(mappedDiagnostics.Any(), diagnosticsByFile.Select(g => g.Key));
+        }
+
+        public async Task<IUpdaterResult> ApplyAsync(IUpgradeContext context, ImmutableArray<RazorCodeDocument> inputs, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var project = GetProjectWithGeneratedCode(context.CurrentProject.Required(), inputs);
+            var originalProject = project;
+            var generatedFilePaths = inputs.Select(GetGeneratedFilePath);
+
+            // Regenerating diagnostics is slow for large projects, but is necessary in between fixing multiple diagnostics
+            // in a single file. To minimize the number of time diagnostics are gathered, fix one diagnostic each
+            // from multiple files before regenerating diagnostics.
+            var diagnostics = await GetDiagnosticsFromTargetFilesAsync(project, context, generatedFilePaths, new LocationAndIDComparer(false), token).ConfigureAwait(false);
+            var diagnosticsCount = diagnostics.Count();
+            while (diagnosticsCount > 0)
+            {
+                // Iterate through the first diagnostic from each document
+                foreach (var diagnostic in diagnostics.GroupBy(d => d.Location.SourceTree?.FilePath).Select(g => g.First()))
+                {
+                    var doc = project.GetDocument(diagnostic.Location.SourceTree);
+                    if (doc is null)
+                    {
+                        continue;
+                    }
+
+                    // Apply the code fix in the generated source
+                    project = await TryFixDiagnosticAsync(diagnostic, doc).ConfigureAwait(false);
+                }
+
+                diagnostics = await GetDiagnosticsFromTargetFilesAsync(project, context, generatedFilePaths, new LocationAndIDComparer(false), token).ConfigureAwait(false);
+                var newDiagnosticCount = diagnostics.Count();
+                if (diagnosticsCount == newDiagnosticCount)
+                {
+                    _logger.LogWarning("Diagnostics could not be fixed as expected. This may be caused by the project being in a bad state (did NuGet packages restore correctly?) or by errors in analyzers or code fix providers related to diagnosticsL {DiagnosticIds}.", string.Join(", ", diagnostics.Select(d => d.Id).Distinct()));
+                    break;
+                }
+                else
+                {
+                    diagnosticsCount = newDiagnosticCount;
+                }
+            }
+
+            // Update cshtml based on changes made to generated source code
+            _logger.LogDebug("Updating Razor documents based on changes made by code fix providers");
```

Comment:
assuming an error pops up, it would be helpful to know which diagnostic we're currently fixing

Reply:
This is fixing lots of different diagnostics at once, though, so it's hard to isolate one.